src/main/java/com/android/volley/toolbox/NetworkUtility.java

```diff
@@ -173,24 +186,22 @@ static void handleException(
                                 responseHeaders);
                 if (statusCode == HttpURLConnection.HTTP_UNAUTHORIZED
                         || statusCode == HttpURLConnection.HTTP_FORBIDDEN) {
-                    attemptRetryOnException("auth", request, new AuthFailureError(networkResponse));
-                } else if (statusCode >= 400 && statusCode <= 499) {
+                    return new RetryInfo("auth", new AuthFailureError(networkResponse));
+                }
+                if (statusCode >= 400 && statusCode <= 499) {
                     // Don't retry other client errors.
                     throw new ClientError(networkResponse);
-                } else if (statusCode >= 500 && statusCode <= 599) {
+                }
+                if (statusCode >= 500 && statusCode <= 599) {
                     if (request.shouldRetryServerErrors()) {
-                        attemptRetryOnException(
-                                "server", request, new ServerError(networkResponse));
-                    } else {
-                        throw new ServerError(networkResponse);
+                        return new RetryInfo("server", new ServerError(networkResponse));
                     }
-                } else {
-                    // 3xx? No reason to retry.
                     throw new ServerError(networkResponse);
```

Comment:
Is this needed? You have it again below

Reply:
Nope, just needed to generalize the comment. Done.